Amqp/src/Akka.Streams.Amqp/AmqpConnectionProvider.cs

```diff
@@ -0,0 +1,419 @@
+        }
+
+        public AmqpDetailsConnectionProvider WithAutomaticRecoveryEnabled(bool automaticRecoveryEnabled)
+        {
+            return new AmqpDetailsConnectionProvider(HostAndPortList, Credentials, VirtualHost, Ssl, RequestedHeartbeat,
+                ConnectionTimeout, HandshakeTimeout, NetworkRecoveryInterval, automaticRecoveryEnabled,
+                TopologyRecoveryEnabled, ConnectionName);
+        }
+
+        public AmqpDetailsConnectionProvider WithTopologyRecoveryEnabled(bool topologyRecoveryEnabled)
+        {
+            return new AmqpDetailsConnectionProvider(HostAndPortList, Credentials, VirtualHost, Ssl, RequestedHeartbeat,
+                ConnectionTimeout, HandshakeTimeout, NetworkRecoveryInterval, AutomaticRecoveryEnabled,
+                topologyRecoveryEnabled, ConnectionName);
+        }
+
+        public AmqpDetailsConnectionProvider WithConnectionName(string connectionName)
+        {
+            return new AmqpDetailsConnectionProvider(HostAndPortList, Credentials, VirtualHost, Ssl, RequestedHeartbeat,
+                ConnectionTimeout, HandshakeTimeout, NetworkRecoveryInterval, AutomaticRecoveryEnabled,
+                TopologyRecoveryEnabled, connectionName);
+        }
+
+        public override IConnection Get()
+        {
+            var factory = new ConnectionFactory();
+            if (Credentials.HasValue)
+            {
+                factory.UserName = Credentials.Value.Username;
+                factory.Password = Credentials.Value.Password;
+            }
+            if (!string.IsNullOrEmpty(VirtualHost))
+                factory.VirtualHost = VirtualHost;
+            if (Ssl != null)
+                factory.Ssl = Ssl;
+            if (AutomaticRecoveryEnabled.HasValue)
+                factory.AutomaticRecoveryEnabled = AutomaticRecoveryEnabled.Value;
+            if (RequestedHeartbeat.HasValue)
+                factory.RequestedHeartbeat = RequestedHeartbeat.Value;
+            if (NetworkRecoveryInterval.HasValue)
+                factory.NetworkRecoveryInterval = NetworkRecoveryInterval.Value;
+            if (TopologyRecoveryEnabled.HasValue)
+                factory.TopologyRecoveryEnabled = TopologyRecoveryEnabled.Value;
+            if (ConnectionTimeout.HasValue)
+                factory.ContinuationTimeout = ConnectionTimeout.Value;
+            if (HandshakeTimeout.HasValue)
+                factory.HandshakeContinuationTimeout = HandshakeTimeout.Value;
+            return factory.CreateConnection(
+                new DefaultEndpointResolver(HostAndPortList.Select(hp => new AmqpTcpEndpoint(hp.host, hp.port))),
+                ConnectionName ?? "");
+        }
+
+        public override string ToString()
+        {
+            return
+                $"AmqpDetailsConnectionProvider(HostAndPortList=({HostAndPortList.Select(x => $"[{x.host}:{x.port}]").Aggregate((left, right) => $"{right}, {left}")}), Credentials={Credentials}, VirtualHost={VirtualHost})";
```

Comment:
It's a bit hard to see what `Aggregate` does here. What about more common `string.Join`?
```csharp
$"AmqpDetailsConnectionProvider(HostAndPortList=({string.Join(", ", HostAndPortList.Select(x => $"[{x.host}:{x.port}]"))}), Credentials={Credentials}, VirtualHost={VirtualHost})";
```